Initial contribution of the Bose Soundtouch binding

kaikreuzer opened the pull request:
This PR replaces https://github.com/openhab/openhab2-addons/pull/2233 and thus ports the work of @tratho and @marvkis to ESH.
I've so far mainly ported the code, but I will also do updates that include my review comments the next days.
I regard this PR to be WIP in the moment, so all others are welcome to create PRs against it with fixes, improvements and new features.

Also-by: Christian Niessner <github-marvkis@christian-niessner.de>
Also-by: Thomas Traunbauer <thomas.traunbauer@gmail.com>
Signed-off-by: Kai Kreuzer <kai@openhab.org>

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/thing/SoundTouch20.xml:
```diff
@@ -0,0 +1,50 @@
+			<parameter name="DEVICE_HOST" type="text" required="true">
+				<label>Host Address</label>
+				<description>The Host / IP Address used for communication to this device.
+                </description>
```
Indentation is misplaced

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/thing/SoundTouch30.xml:
```diff
@@ -0,0 +1,50 @@
+			<parameter name="DEVICE_HOST" type="text" required="true">
+				<label>Host Address</label>
+				<description>The Host / IP Address used for communication to this device.
+                </description>
```
Indentation

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/thing/SoundTouch300.xml:
```diff
@@ -0,0 +1,49 @@
+			<parameter name="DEVICE_HOST" type="text" required="true">
+				<label>Host Address</label>
+				<description>The Host / IP Address used for communication to this device.
+                </description>
```
indentation

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/thing/SoundTouchSA5Amplifier.xml:
```diff
@@ -0,0 +1,50 @@
+			<parameter name="DEVICE_HOST" type="text" required="true">
+				<label>Host Address</label>
+				<description>The Host / IP Address used for communication to this device.
+                </description>
```
indentation

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/thing/SoundTouchWirelessLinkAdapter.xml:
```diff
@@ -0,0 +1,49 @@
+			<parameter name="DEVICE_HOST" type="text" required="true">
+				<label>Host Address</label>
+				<description>The Host / IP Address used for communication to this device.
+                </description>
```
indentation

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/thing/UnknownSoundTouch.xml:
```diff
@@ -0,0 +1,50 @@
+			<parameter name="DEVICE_HOST" type="text" required="true">
+				<label>Host Address</label>
+				<description>The Host / IP Address used for communication to this device.
+                </description>
```
indentation

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/ESH-INF/thing/WaveSoundTouchMusicSystemIV.xml:
```diff
@@ -0,0 +1,50 @@
+			<parameter name="DEVICE_HOST" type="text" required="true">
+				<label>Host Address</label>
+				<description>The Host / IP Address used for communication to this device.
+                </description>
```
indentation

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/BoseSoundTouchBindingConstants.java:
```diff
@@ -0,0 +1,85 @@
+    public static final String BINDING_ID = "bosesoundtouch";
+
+    // List of all Thing Type UIDs
+    public final static ThingTypeUID BST_UNKNOWN_THING_TYPE_UID = new ThingTypeUID(BINDING_ID, "device");
```
static final instead if final static. applies to all the constants.

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/internal/ContentItem.java:
```diff
@@ -0,0 +1,259 @@
+     * @return the operation Mode, depending on the stats that are set
+     */
+    public OperationModeType getOperationMode() {
+        OperationModeType operationMode = OperationModeType.OTHER;
```
I believe we can move the whole method to OperationModeType

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/internal/ContentItem.java:
```diff
@@ -0,0 +1,259 @@
+
+    @Override
+    public String toString() {
+        // if (presetID >= 1 && presetID <= 6) {
```
we could remove the unnecessary comments

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/internal/discovery/DiscoveryUtil.java:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @author Thomas Traunbauer - Initial contribution
+ */
+public class DiscoveryUtil {
```
We could add a private constructor

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/internal/discovery/SoundTouchDiscoveryParticipant.java:
```diff
@@ -0,0 +1,169 @@
+            ThingTypeUID typeUID = getThingTypeUID(info);
+            if (typeUID != null) {
+                if (info.getType() != null) {
+                    if (info.getType().equals(getServiceType())) {
```
These 3 if conditions can be merged together

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/internal/discovery/SoundTouchDiscoveryParticipant.java:
```diff
@@ -0,0 +1,169 @@
+
+    @Override
+    public String getServiceType() {
+        return "_soundtouch._tcp.local.";
```
can be extracted as constant

amitjoy commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/internal/CommandExecutor.java:
```diff
@@ -0,0 +1,710 @@
+                Collection<ContentItem> listOfPresets = presetContainer.getAllPresets();
+                for (ContentItem ps : listOfPresets) {
+                    if (ps.isPresetable()) {
+                        if (ps.getLocation().equals(contentItem.getLocation())) {
```
These 2 if conditons can be merged

kaikreuzer commented:
@amitjoy Thanks for your comments, but please note that I marked this WIP and not yet in a state to be reviewed. So as mentioned above, if you want to help improving the code, please directly create a PR against this branch. Thanks!

amitjoy commented:
@kaikreuzer Sounds like a plan 😊

tratho commented:
Hi @kaikreuzer nice to see the PR is created.
One simple question. What does this WIP mean?

amitjoy commented:
@tratho **W**ork **I**n **P**rogress

marvkis commented:
Hi, i'm also very happy to see this PR here ;) Thanks a lot @kaikreuzer for all of it.
Is there already a plan for the next step steps? Maybe I find some time to contribute a few things.

kaikreuzer commented:
> Is there already a plan for the next step steps?

Yes, see above: " I will also do updates that include my review comments the next days."
My goal would be some clean up and potentially a slight feature reduction to get a first version merged soon - once this is there, it will be much easier to work on individual features and extensions in separate PRs.

marvkis commented:
Okay, then I'll wait for your updates and jump in after that. If there are any questions don't hesitate to ask ;)

kaikreuzer commented:
@marvkis & @tratho: Short update: As you can see at https://github.com/eclipse/smarthome/pull/5397/commits, I have done a couple of changes - I tried to keep them in separate commits so that it is easy to track what I have changed. Feel free to comment and veto.

The main functional changes so far were:
- removal of the "preset control" channel as the Player item is not really a good match here
- changed the artwork channel from a String (with url) to an Image, so that it works in all UIs (also Paper UI)
- changed the preset container to use standard storage mechanism instead of serialising the content to a json file by itself.

I am not yet fully done with my review and changes, but have already covered quite a bit.
In general, it works quite nice, so thumbs up 👍 .

Two small questions:

1. Whenever I use the Bose app and change something, I get warnings and also debug with "Unexpected text data..." in my logs - do you have those as well or what could be wrong here?

```
2018-04-16 23:42:38.449 [DEBUG] [.s.b.b.h.BoseSoundTouchHandler:403  ] - Demo Speaker: onWebSocketText('<userActivityUpdate deviceID="7C010A8E1A14" />')
2018-04-16 23:42:38.450 [WARN ] [o.e.s.b.b.i.XMLResponseHandler:109  ] - Demo Speaker: Unhandled XML entity during INIT: 'userActivityUpdate'
2018-04-16 23:42:38.471 [DEBUG] [.s.b.b.h.BoseSoundTouchHandler:403  ] - Demo Speaker: onWebSocketText('<updates deviceID="7C010A8E1A14"><volumeUpdated><volume><targetvolume>13</targetvolume><actualvolume>13</actualvolume><muteenabled>false</muteenabled></volume></volumeUpdated></updates>')
2018-04-16 23:42:38.472 [WARN ] [o.e.s.b.b.i.XMLResponseHandler:325  ] - Demo Speaker: Unhandled XML entity during VolumeUpdated: 'volume'
2018-04-16 23:42:38.534 [DEBUG] [.s.b.b.h.BoseSoundTouchHandler:403  ] - Demo Speaker: onWebSocketText('<?xml version="1.0" encoding="UTF-8" ?><msg><header deviceID="7C010A8E1A14" url="volume" method="GET"><request requestID="0" msgType="RESPONSE"><info type="new" /></request></header><body><volume deviceID="7C010A8E1A14"><targetvolume>13</targetvolume><actualvolume>13</actualvolume><muteenabled>false</muteenabled></volume></body></msg>')
2018-04-16 23:42:38.535 [DEBUG] [o.e.s.b.b.i.XMLResponseHandler:469  ] - Demo Speaker: Unexpected text data during VolumeTarget: '13'
```

2. It seems that the REFRESH command is not yet supported. This is sent to the handler, if channels get linked, so it should not be ignored (otherwise newly linked channels do not have a state). What is the best way to retrieve the latest state of a specific channel?

tratho commented:
@kaikreuzer to your questions
1. The API seems to change with every update. But normally it seems to be backward compatible. So it just gets extended. The Perhaps the warnings should be debug messages to. The main benefit is to see if the API is extended in a way, that could be used in the binding. But this part is nearly 100% from @marvkis propably he can tell you more about that.

2. No the REFRESH command is not supported yet. The problem is i do not have an idea how to get the latest state. So there is still a //TODO

tratho commented:
all other changes seems to be good.
I know the prestControl as player is not the best. But is there an other item, that only supports forward/backward?

marvkis commented:
Hi @kaikreuzer, to your questions:
1. It is like @tratho  already stated out. Especially in the past it happened quite often that the api was changed during firmware updates and so I decided to have the parser stating unexpected data with warning messages to see it quite quickly. I still don't use the official API doc's, but from discussions with other developers knowing the API docs I know that the doc's are not covering these changes in time so we have to do this in a empirical way ;) But the API seems to be quite stable now.
Having a look on the logs you posted: Might it be that you did a "manual" volume change on the speaker or remote? For me the `<userActivityUpdate/>` might be a notification that some user activity occurred. Now we have to decide if this message is relevant for us, i.e. we want to trigger a 'user activitiy channel' or if we want to ignore this.

marvkis commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/internal/XMLResponseProcessor.java:
```diff
@@ -0,0 +1,148 @@
+        updatesMap.put("volumeUpdated", XMLHandlerState.MsgBody);
+        updatesMap.put("zoneUpdated", XMLHandlerState.ZoneUpdated); // just notifies but dosn't provide details
+        updatesMap.put("bassUpdated", XMLHandlerState.BassUpdated);
+        updatesMap.put("volumeUpdated", XMLHandlerState.VolumeUpdated);
```
This seems the point with `volumeUpdated`. It is put here into the map two times - one for MsgBody in line 87 (that would work for the message as we see it here) and one time for 'VolumeUpdated' in line 90 overriding the value from line 87. I traced the `VolumeUpdated` part and it seems in the past only the `<volumeUpdated/>` tag without the `<volume>` data was sent and so `<volumeUpdated/>` triggered a fetch for the current volume. Seeing the message from the log it seems we could clean up the entire `VolumeUpdated` path...

triller-telekom commented:
@marvkis Regarding your REFRESH command question:

> Is there an easy way to trigger a REFRESH for an channel manually (via the CLI?) 

You can link an item to a channel and afterwards use the console to send the REFRESH via

```
smarthome send ItemName REFRESH
```

kaikreuzer commented:
> I know the prestControl as player is not the best. But is there an other item, that only supports forward/backward?

No, there is no item type for simply forward/backward commands. I think the most straight forward way is to use the `keyCode` channel and use mapping information in a sitemap to render the buttons that one wishes to use (or to change the preset through a rule). Hence I think we are fine with the channels as they are now.

kaikreuzer commented:
All, I have done further clean up to bring this in a state to create a CQ for the Eclipse IP team.
A major change I had to do - as it very much convoluted the architecture by requiring dependencies between different handlers and factories - is the removal of the zone info/add/remove support. My suggestion is to rather check if we can introduce rule actions that allow the management of zones - such a feature is anyhow not a good fit for Thing channels.

Please refrain from any changes for the moment - we will merge this PR as soon as we have approval from the IP team and can then follow up with individual PRs afterwards.

kaikreuzer commented:
FTR, I have created CQ 16146.
As I had to squash the commits on this branch, I kept the old commit list for reference of my latest changes at https://github.com/eclipse/smarthome/compare/master...kaikreuzer:bosesoundtouch_detail.

tratho commented:
@kaikreuzer nice to see you finished the progress.
One thing is very sad for me. You removed of the zone handling.
I need that very much. How do these Rule Actions work you mentioned?

ivivanov-bg commented on extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/internal/discovery/SoundTouchDiscoveryParticipant.java:
```diff
@@ -0,0 +1,170 @@
+            }
+
+            properties.put(DEVICE_PARAMETER_HOST, addrs[0].getHostAddress());
+            if (getMacAddress(info) != null) {
```
I guess an error should be raised if the MAC address is not available (or make it not required in the config.xml)

kaikreuzer replied:
Not necessarily - it would still be a valid discovery result, but the user would have to figure out the MAC by himself then (and enter it in the config).

kaikreuzer commented:
> How do these Rule Actions work you mentioned?

This isn't yet fully worked out, but a feature that is put on the [ESH backlog](https://github.com/eclipse/smarthome/projects/3) to indicate that it is an important topic to work on next. https://github.com/openhab/openhab2-addons/pull/2926 is the first prototype that should be used for discussion.

kaikreuzer commented:
CQ has been approved!
@maggu2810, @sjka or @htreu - please merge as is.

zr2002 commented:
> One thing is very sad for me. You removed of the zone handling.
I need that very much. How do these Rule Actions work you mentioned?

@kaikreuzer thank you for your great job  
I am also using the zone feature in a habpanel widget and I hope we will find a good alternative
